### ISSUE TITLE
local install of gem build issue

### ISSUE THREAD

**Issue by rinkevichjm** [shoes/shoes4#1165]
C:\Users\james.rinkevich\Downloads>java -jar jruby-complete-9.0.1.0.jar -S gem install -p https://nmciproxyb1secure:8443 --user-install shoes --pre
Successfully installed nokogiri-1.6.4.1-java
Fetching: shoes-highlighter-1.0.1.gem (100%)
Successfully installed shoes-highlighter-1.0.1
Fetching: shoes-manual-4.0.0.gem (100%)
Successfully installed shoes-manual-4.0.0
Fetching: rubyzip-1.1.7.gem (100%)
Successfully installed rubyzip-1.1.7
Fetching: plist-3.1.0.gem (100%)
Successfully installed plist-3.1.0
Fetching: jruby-rack-1.1.19.gem (100%)
Successfully installed jruby-rack-1.1.19
Fetching: jruby-jars-1.7.22.gem (100%)
Successfully installed jruby-jars-1.7.22
Fetching: warbler-1.4.9.gem (100%)
Successfully installed warbler-1.4.9
Fetching: furoshiki-0.3.1.gem (100%)
Successfully installed furoshiki-0.3.1
Fetching: shoes-core-4.0.0.pre4.gem (100%)
Building native extensions.  This could take a while...
ERROR:  Error installing shoes:
        ERROR: Failed to build gem native extension.

    java -cp ;C:\Users\james.rinkevich\Downloads\jruby-complete-9.0.1.0.jar org.jruby.Main -rubygems uri:classloader:/META-INF/jruby.home/lib/rub
y/gems/shared/gems/rake-10.1.0/bin/rake RUBYARCHDIR=C:/Users/james.rinkevich/.gem/jruby/2.2.0/extensions/universal-java-1.7/2.2.0/shoes-core-4.0.
0.pre4 RUBYLIBDIR=C:/Users/james.rinkevich/.gem/jruby/2.2.0/extensions/universal-java-1.7/2.2.0/shoes-core-4.0.0.pre4
rake aborted!
No such file or directory - uri:classloader:/META-INF/jruby.home/bin/shoes.bat
org/jruby/RubyFile.java:342:in `initialize'
org/jruby/RubyIO.java:1124:in `open'
org/jruby/RubyIO.java:1126:in `open'
C:/Users/james.rinkevich/.gem/jruby/2.2.0/gems/shoes-core-4.0.0.pre4/ext/install/rakefile:28:in `block in (root)'
org/jruby/RubyProc.java:308:in `call'
org/jruby/RubyArray.java:1560:in `each'
org/jruby/RubyArray.java:1560:in `each'
org/jruby/RubyArray.java:1560:in `each'
Tasks: TOP => default => install_script
(See full trace by running task with --trace)

rake failed, exit code 1

Gem files will remain installed in C:/Users/james.rinkevich/.gem/jruby/2.2.0/gems/shoes-core-4.0.0.pre4 for inspection.
Results logged to C:/Users/james.rinkevich/.gem/jruby/2.2.0/extensions/universal-java-1.7/2.2.0/shoes-core-4.0.0.pre4/gem_make.out

**Comment by jasonrclark** [comment]
Thanks for the bug report!

I've never seen running JRuby via `java -jar jruby-complete-9.0.1.0.jar`. I suspect that's the problem--the shoes gem installation (which has some custom tests) is getting a path relative to the internals of `jruby-complete-9.0.1.0.jar` instead of the outside file-system.

Is there a reason you aren't running `jruby` more directly?

**Comment by rinkevichjm** [comment]
Our IT department alows us to run java apps but anything else requires a special review

**Comment by PragTob** [comment]
I see, thanks for the report. We'll see to make shoes run in that environment (no promises on times though)

**Comment by rinkevichjm** [comment]
It might be related to JRUBY_HOME not being set.

**Comment by jasonrclark** [comment]
Dug in a little, and the problem is this line in our custom installation code: https://github.com/shoes/shoes4/blob/master/shoes-core/ext/install/Rakefile#L22. In particular, that uses the `__FILE__` to try and determine where the gem's unpacked files are on disk. This works in most cases, but when running out of the jar, the path is mangled.

Although I haven't looked yet, it should be possible to determine where Rubygems has unpacked the files to on disk more directly and use that concrete path, which would fix this. Given that's a pretty clear course, I've set this for 4.0.0-rc1 and we should get this fixed before the full 4.0 release.

**Comment by jasonrclark** [comment]
Oh, and for those (like me!) who didn't know about `jruby-complete`, you can just get the jar file directly from the [JRuby download page!](http://jruby.org/download) It easily duplicates the problem @rinkevichjm reported once you've got that piece downloaded.

**Comment by jasonrclark** [comment]
Hey @rinkevichjm! Ended up taking a look at this over on PR #1168 and I *might* have it covered there.

I don't know what capacity you have to test things out, but if you wanted to take a look on that PR I would greatly appreciate it. To do that, you'll need to:

* Clone the shoes repo and get onto the `jclark/jruby-complete` brach
* Run `rake build:all` to build the Shoes gems
* Run the gem installation command like you were doing before, but instead of passing `shoes --pre`, you'll pass the file paths where the prior command built the gems to:
  * `shoes-core/pkg/shoes-core-4.0.0.pre4.gem`
  * `shoes-package/pkg/shoes-package-4.0.0.pre4.gem`
  * `shoes-swt/pkg/shoes-swt-4.0.0.pre4.gem`

It also appears that even once I got the gem properly installed, I couldn't use it the normal way of invoking `shoes my_shoes_app.rb`, but there is a workaround. Put the following in the top of your Shoes app file:

```
require 'shoes'
```

And then when you run, you'll want to do something more like this:

```
java -jar jruby-complete-9.0.1.0.jar -S my_shoes_app.rb
```

This directly executes your Ruby file, but the require line we added should get the proper version of Shoes loaded up.

Sorry for the crazy-long set of steps! Turns out that testing installation before something gets pushed to Rubygems is a bit rough, but hopefully I've captured the necessary pieces here. Thanks again!

**Comment by rinkevichjm** [comment]
I am out of the office next week at training so I don't know if I can test until the week after. I have managed to work around it and get shoes running. It is nice.

**Comment by PragTob** [comment]
Thanks for the praise :) Have fun at the training, would be great if you could check it out either way when you get back - really want that to work :)

**Comment by jasonrclark** [comment]
New changes are available on https://github.com/shoes/shoes4/pull/1168 that I think finally deal with all the issues. Managed to get it tested on Windows, Mac and Linux, but would appreciate a second set of eyes if anyone has a few moment to check it.

**Comment by PragTob** [comment]
@jasonrclark :heart: :heart: :heart_eyes: :heart_eyes: :clap: :pray:

**Comment by PragTob** [comment]
Fixed through #1168 ! Thanks for the report, I'm just gonna go ahead and say we'll make a release this year to include the changes (ideally before Christmas, but knowing the time of @jasonrclark and I it might be in the time gap between Christmans and new year's)
